src/batch/src/executor2/order_by.rs

```diff
@@ -63,186 +56,74 @@ impl BoxedExecutorBuilder for OrderByExecutor {
+impl OrderByExecutor2 {
+    #[try_stream(boxed, ok = DataChunk, error = RwError)]
+    async fn do_execute(self: Box<Self>) {
+        let data_types = self.schema().data_types();
+        let mut chunks = vec![];
+        let mut stream = self.child.execute();
+        while let Some(chunk) = stream.next().await {
+            chunks.push(chunk?);
         }
-
-        self.collect_child_data().await?;
-
-        self.child.close().await?;
-        Ok(())
-    }
-
-    async fn next(&mut self) -> Result<Option<DataChunk>> {
-        let data_types = self
-            .schema()
-            .fields()
+        let mut indexes = gen_index_array(&chunks);
+        let comparators = self.order_pairs;
+        indexes.sort_unstable_by(|row1, row2| cmp(row1, row2, &comparators));
+        // build chunk by the new order
+        let rows = indexes
             .iter()
-            .map(|f| f.data_type.clone())
-            .collect_vec();
-        let mut builders = data_types
-            .iter()
-            .map(|t| t.create_array_builder(K_PROCESSING_WINDOW_SIZE))
-            .collect::<Result<Vec<ArrayBuilderImpl>>>()?;
-        let mut chunk_size = 0usize;
-        while !self.min_heap.is_empty() && chunk_size < K_PROCESSING_WINDOW_SIZE {
-            let top = self.min_heap.pop().unwrap();
-            for (idx, builder) in builders.iter_mut().enumerate() {
-                let chunk_arr = self.chunks[top.chunk_idx].column_at(idx).array();
-                let chunk_arr = chunk_arr.as_ref();
-                macro_rules! gen_match {
-                    ($b: ident, $a: ident, [$( $tt: ident), *]) => {
-                        match ($b, $a) {
-                            $((ArrayBuilderImpl::$tt($b), ArrayImpl::$tt($a)) => Ok($b.append($a.value_at(top.elem_idx))),)*
-                                _ => Err(InternalError(String::from("Unmatched array and array builder types"))),
-                        }?
-                    }
-                }
-                let _ = gen_match!(
-                    builder,
-                    chunk_arr,
-                    [
-                        Int16,
-                        Int32,
-                        Int64,
-                        Float32,
-                        Float64,
-                        Utf8,
-                        Bool,
-                        Decimal,
-                        Interval,
-                        NaiveDate,
-                        NaiveTime,
-                        NaiveDateTime
-                    ]
-                );
-            }
-            chunk_size += 1;
-            self.push_heap_for_chunk(top.chunk_idx);
-        }
-        if chunk_size == 0 {
-            return Ok(None);
+            .map(|row_ref| row_ref.to_owned_row())
+            .collect::<Vec<_>>();
+        let data_chunk = DataChunk::from_rows(&rows, &data_types)?;
+        let chunks = DataChunk::rechunk(&[Arc::new(data_chunk)], self.chunk_size)?;
```

Comment:
These two lines will consume 3 times memory than original version.

Reply:
> Please keep original version. The encodings was used to vectorized comparison.

ok, I will use the original version